### PR TITLE
r/persistent_volume - add support for CSI

### PR DESCRIPTION
**Pull request by DrFaust92**
### Acceptance tests
- [x] Have you added an acceptance test for the functionality being added?
- [x] Have you run the acceptance tests on this branch? (If so, please include the test log in a gist)

```
--- PASS: TestAccKubernetesPersistentVolume_csi_basic (3.65s)
```

### References

Relates #296

### CODE REVIEW

File: kubernetes/schema_volume_source.go [hunk 1]
```diff
@@ -218,6 +218,41 @@ func commonVolumeSources() map[string]*schema.Schema {
...
+			Elem: &schema.Resource{
+				Schema: map[string]*schema.Schema{
+					"driver": {
+						Type:        schema.TypeString,
+						Description: "the name of the volume driver to use. More info: https://kubernetes.io/docs/concepts/storage/volumes/#csi",
+						Required:    true,
+					},
+					"volume_handle": {
+						Type:        schema.TypeString,
+						Description: "A map that specifies static properties of a volume. More info: https://kubernetes.io/docs/concepts/storage/volumes/#csi",
```
Comment (benosman) [inline kubernetes/schema_volume_source.go:235]: This description is incorrect - should refer to a string like:
`A string value that uniquely identifies the volume. `

Comment (DrFaust92) [reply kubernetes/schema_volume_source.go]: Fixed

**Comment by benosman** [comment]
I'm just an end user, but I tried out this pull request and with my initial testing, so far it's working as it should for me.

**Comment by DrFaust92** [comment]
Thanks for the input @benosman

File: kubernetes/structure_persistent_volume_spec.go [hunk 2]
```diff
@@ -1518,6 +1559,28 @@ func patchPersistentVolumeSource(pathPrefix, prefix string, d *schema.ResourceDa
 		}
 	}
 
+	if d.HasChange(prefix + "csi") {
+		oldIn, newIn := d.GetChange(prefix + "csi")
+		oldV, oldOk := oldIn.([]interface{})
+		newV, newOk := newIn.([]interface{})
+
+		if newOk && len(newV) > 0 {
+			if oldOk && len(oldV) > 0 {
```
Comment (jrhouston) [inline kubernetes/structure_persistent_volume_spec.go:1568]: I think we could clean the duplication here by setting 

```go
path := pathPrefix + "/csi"
value := expandCSIPersistentDiskVolumeSource(newV)
```

before the conditional, and then just supply them when constructing AddOperation/ReplaceOperation

File: kubernetes/schema_volume_source.go [hunk 3]
```diff
@@ -218,6 +218,41 @@ func commonVolumeSources() map[string]*schema.Schema {
 				},
 			},
 		},
+		"csi": {
+			Type:        schema.TypeList,
+			Description: "Represents a CSI Volume. More info: http://kubernetes.io/docs/user-guide/volumes#csi",
+			Optional:    true,
+			MaxItems:    1,
+			Elem: &schema.Resource{
```
Comment (jrhouston) [inline kubernetes/schema_volume_source.go:226]: I see in the [API documentation](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#csipersistentvolumesource-v1-core) for this resource there are several additional attributes for referencing secrets - is there a reason we haven't added them here?

File: kubernetes/structure_persistent_volume_spec.go (hunk 2 above)
Comment (DrFaust92) [reply kubernetes/structure_persistent_volume_spec.go:1568]: should i just do this for all other as well? maybe extract to a more general function.
id addressed the changes you mentioned, i can open another PR for a more general refactor

File: kubernetes/schema_volume_source.go (hunk 3 above)
Comment (DrFaust92) [reply kubernetes/schema_volume_source.go:212]: they are a bit tricky but i added them.
while the secret_ref state that `namespace` is optional. the api call fails if it not set. i extracted the `secret_ref` block to a common func (maybe better to extract to provider wide later)
i kept the argument optional as stated in the docs (to also make the block common). its a non breaking change as the namespace is also computed.

**Comment by DrFaust92** [comment]
addressed comments + some "minor" changes.

tests output:
```
--- PASS: TestAccKubernetesPersistentVolume_csi_basic (5.32s)
--- PASS: TestAccKubernetesPersistentVolume_csi_importBasic (2.89s)
--- PASS: TestAccKubernetesPersistentVolume_csi_secrets (2.69s)
```

**Comment by DrFaust92** [comment]
@jrhouston, polite nudge

**Pull request merged**